include/vsag/index_feature.h

```diff
@@ -0,0 +1,57 @@
+
+// Copyright 2024-present the vsag project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+namespace vsag {
+
+enum class IndexFeature {
+    NEED_TRAIN,
+    SUPPORT_BUILD,
+    SUPPORT_ADD,
+    SUPPORT_ADD_FROM_EMPTY,
+    SUPPORT_KNN_SEARCH,
+    SUPPORT_KNN_SEARCH_WITH_ID_FILTER,
+    SUPPORT_RANGE_SEARCH,
+    SUPPORT_RANGE_SEARCH_WITH_ID_FILTER,
+    SUPPORT_DELETE_BY_ID,
+    SUPPORT_BATCH_SEARCH,
+    SUPPORT_METRIC_TYPE_L2,
+    SUPPORT_METRIC_TYPE_INNER_PRODUCT,
+    SUPPORT_METRIC_TYPE_COSINE,
+    SUPPORT_SERIALIZE_FILE,
+    SUPPORT_SERIALIZE_BINARY_SET,
+    SUPPORT_DESERIALIZE_FILE,
+    SUPPORT_DESERIALIZE_BINARY_SET,
+    SUPPORT_DESERIALIZE_READER_SET,
+    SUPPORT_RESET,
+    SUPPORT_FEEDBACK,
+    SUPPORT_CAL_DISTANCE_BY_ID,
+
+    SUPPORT_TRAIN_WITH_MULTI_THREAD,
+    SUPPORT_BUILD_WITH_MULTI_THREAD,
+    SUPPORT_BATCH_ADD_WITH_MULTI_THREAD,
+    SUPPORT_SEARCH_ONE_WITH_MULTI_THREAD,
+    SUPPORT_BATCH_SEARCH_WITH_MULTI_THREAD,
+
+    SUPPORT_ADD_CONCURRENT,
+    SUPPORT_SEARCH_CONCURRENT,
+    SUPPORT_DELETE_CONCURRENT,
+    SUPPORT_ADD_SEARCH_CONCURRENT,
+    SUPPORT_ADD_DELETE_CONCURRENT,
+    SUPPORT_SEARCH_DELETE_CONCURRENT,
+    SUPPORT_ADD_SEARCH_DELETE_CONCURRENT,
```

Comment:
should we make all of them true in the future?

Reply:
not for all, some index may static without add and search concurrently. some feature can be enable by pass different params to factory. but may bring some performance loss. so user can disable these feature, just avoiding the concurrent between such behaviors in their own process with vsag